src/resolver.rs

```diff
@@ -188,14 +181,28 @@ pub struct CriteriaFailureSet {
     pub all: CriteriaSet,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
+pub enum CriteriaNamespace {
+    Local,
+    Foreign(ImportName),
+}
+
+#[derive(Debug, Clone)]
+pub struct CriteriaInfo {
+    pub namespace: CriteriaNamespace,
+    pub raw_name: CriteriaName,
+    pub namespaced_name: String,
+    pub entry: CriteriaEntry,
```

Comment:
It seems like it might be nice to include the implies CriteriaSet in here too, rather than just holding the single CriteriaEntry. We could have a `direct_implies` set which is populated immediately from the entries implies relationship, and then fill in `implies` afterward recursively like we're already doing.

I don't know if we really want the full CriteriaEntry here because the only extra info it has over that right now is descriptions, which could be large and IIRC the resolver never uses internally so aren't important to hold onto (though I could be forgetting something there).

Reply:
this is a good point. i think holding onto the entry was largely an artifact of the long-past days where CriteriaMapper tried really hard to not do any clones and just hold onto borrowed things, but builtins got in the way of this, so that hasn't been the case for a while.